### PR TITLE
SONARTEXT-48 Extend Post-Filter to include heuristic Filter

### PR DESCRIPTION
**Pull request by petertrr**

### CODE REVIEW

File: sonar-text-plugin/src/test/java/org/sonar/plugins/secrets/api/HeuristicsTest.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+class HeuristicsTest {
+    @ParameterizedTest
+    @CsvSource({
+            "/home/user, true",
+            "/home/username, false",
+            "C:\\Users\\User, true",
+            "asecretstr/ng, false"
+    })
+    void shouldDeterminePath(String input, boolean isPath) {
+        Assertions.assertThat(Heuristics.isPath(input)).isEqualTo(isPath);
```
Comment (jonas-wielage-sonarsource) [inline sonar-text-plugin/src/test/java/org/sonar/plugins/secrets/api/HeuristicsTest.java:36]: Should be a static import

File: sonar-text-plugin/src/main/java/org/sonar/plugins/secrets/api/Heuristics.java
Comment (jonas-wielage-sonarsource) [inline sonar-text-plugin/src/main/java/org/sonar/plugins/secrets/api/Heuristics.java:1]: Please format this class, as there are 8 code smells related to it.

File: sonar-text-plugin/src/main/java/org/sonar/plugins/secrets/api/Heuristics.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+                default:
+                    LOG.warn("Heuristic with the name `{}` is not supported", h);
+                    return false;
+            }
+        });
+    }
+
+    public static boolean isPath(String input) {
+        long fileSeparatorCount = input.chars().filter(c -> c == '/' || c == '\\').count();
+        return fileSeparatorCount >= 2 && fileSeparatorCount * 1. / input.length() > 0.15;
```
Comment (jonas-wielage-sonarsource) [inline sonar-text-plugin/src/main/java/org/sonar/plugins/secrets/api/Heuristics.java:48]: `0.15` and `2` should be variables at top level. Also please use parenthesis to make calculation more readable.

File: sonar-text-plugin/src/main/java/org/sonar/plugins/secrets/api/Heuristics.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    private static final Logger LOG = Loggers.get(Heuristics.class);
+
+    public static boolean matchesHeuristics(String candidateSecret, List<String> heuristics) {
+        return heuristics.stream().anyMatch(h -> {
+            switch (h) {
+                case "path": return isPath(candidateSecret);
+                case "uri": return isUri(candidateSecret);
+                default:
+                    LOG.warn("Heuristic with the name `{}` is not supported", h);
+                    return false;
```
Comment (jonas-wielage-sonarsource) [inline sonar-text-plugin/src/main/java/org/sonar/plugins/secrets/api/Heuristics.java:41]: Good default

File: sonar-text-plugin/src/main/java/org/sonar/plugins/secrets/api/Heuristics.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+public class Heuristics {
+    private Heuristics() {}
+
+    private static final Pattern uriPattern = Pattern.compile("^(https?|ftps?|file|smtp|imap)://.*$");
+
+    private static final Logger LOG = Loggers.get(Heuristics.class);
+
+    public static boolean matchesHeuristics(String candidateSecret, List<String> heuristics) {
+        return heuristics.stream().anyMatch(h -> {
+            switch (h) {
```
Comment (jonas-wielage-sonarsource) [inline sonar-text-plugin/src/main/java/org/sonar/plugins/secrets/api/Heuristics.java:36]: We don't use single character variable names ;)
Also I like if else more, but should be fine.

File: sonar-text-plugin/src/test/java/org/sonar/plugins/secrets/api/HeuristicsTest.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.plugins.secrets.api;
+
+import java.util.Arrays;
+import org.assertj.core.api.Assertions;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.CsvSource;
+
+class HeuristicsTest {
```
Comment (jonas-wielage-sonarsource) [inline sonar-text-plugin/src/test/java/org/sonar/plugins/secrets/api/HeuristicsTest.java:27]: There should be a test when the `default` path is taken.

**Review by jonas-wielage-sonarsource — APPROVED** [review]
LGTM! Please see my suggestions. Also I would like a test in postFilterSpec.yaml to show that it really works based on the heuristics.

File: sonar-text-plugin/src/main/java/org/sonar/plugins/secrets/api/Heuristics.java (hunk 4 above)
Comment (petertrr) [reply sonar-text-plugin/src/main/java/org/sonar/plugins/secrets/api/Heuristics.java:36]: Just thought that there might be more heuristics in the future, and `switch` would make more sense in that case

Comment (jonas-wielage-sonarsource) [reply sonar-text-plugin/src/main/java/org/sonar/plugins/secrets/api/Heuristics.java]: Sounds good!

**Pull request merged**